File: tree/993.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isCousins(root *TreeNode, x int, y int) bool {
    if root == nil {
        return false
    }
    parentX, depthX := findNode(root, nil, 0, x)
    parentY, depthY := findNode(root, nil, 0, y)
    if parentX != parentY && depthX == depthY {
        return true
    }else {
        return false
    }
    
}

func findNode(root *TreeNode, parent *TreeNode, depth int, val int) (retp *TreeNode, retd int) {
    if root.Val == val {
        retp = parent
        retd = depth
    }else{
        if root.Left != nil {
            retp, retd = findNode(root.Left, root, depth+1, val)   
            
        }
        if retp != nil {
            return retp, retd
        }        
        if root.Right != nil {
            retp, retd = findNode(root.Right, root, depth+1, val)    
        }
    }
    return retp, retd
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isCousins(root *TreeNode, x int, y int) bool {
    var depthx, depthy int
    var prex, prey *TreeNode
    dfs(root, nil, 0, x, y, &depthx, &depthy, &prex, &prey)
    if depthx == depthy && prex != prey {
        return true
    }
    return false
}

func dfs(root *TreeNode, pre *TreeNode, depth int, x int, y int, depthx *int, depthy *int, prex **TreeNode, prey **TreeNode) {
    if root == nil {
        return
    }
    if root.Val == x {
        *depthx = depth
        *prex = pre
        return
    }else if root.Val == y {
        *depthy = depth
        *prey = pre
        return
    }else {
        dfs(root.Left, root, depth+1, x, y, depthx, depthy, prex, prey)
        dfs(root.Right, root, depth+1, x, y, depthx, depthy, prex, prey)
    }
}

File: dfs/993.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isCousins(TreeNode* root, int x, int y) {
        int depth_x = 0, depth_y = 0, father_x = -1, father_y = -1;
        recur(root, x, depth_x, father_x, 0);
        recur(root, y, depth_y, father_y, 0);
        if (father_x == -1 || father_y == -1)
            return false;
        return father_x != father_y && depth_x == depth_y;
    }
private:
    void recur(TreeNode* root, int target, int& depth, int& father, int cur) {
        if (!root)
            return;
        if (root->left && root->left->val == target) {
            depth = cur;
            father = root->val;
            return;
        } 
        if (root->right && root->right->val == target) {
            depth = cur;
            father = root->val;
            return;
        }
        recur(root->left, target, depth, father, cur+1);
        recur(root->right, target, depth, father, cur+1);
        return;
    }
};
